### ISSUE TITLE
FSharp tests + Altcover crashes with Stack overflow (core 3.0 preview)

### ISSUE THREAD

**Issue by eternaldensity** [SteveGilham/altcover#73]
Edit before posting: this has a happy ending.  The problem occurs on a core 3.0 preview but is resolved when I added a global.json pinning it to 2.2.402.

This happened to me with some small toy xunit tests of another F# assembly, but it still happened when I removed all but the initial 'AssertTrue(true)' test that's provided, so I also tried out another project which has Expecto tests (a specifically F# testing framework) and it failed with the same error.

Then I tried generating a fresh F# xunit project and adding Altcover to it and it had the same problem.

The error is: The active test run was aborted. Reason: Test host process crashed : Stack overflow.
It occurs when executing "dotnet test /p:Altcover=true", after starting test execution and finding the test.  (the test is just AssertTrue(true).)

"dotnet test" works and completes just fine.

In full:

```
PS C:\dev> dotnet new xunit -lang F# -n fxunit
The template "xUnit Test Project" was created successfully.

Processing post-creation actions...
Running 'dotnet restore' on fxunit\fxunit.fsproj...
  Restore completed in 544.88 ms for C:\dev\fxunit\fxunit.fsproj.

Restore succeeded.

PS C:\dev> cd .\fxunit\
PS C:\dev\fxunit> dotnet add package Altcover
  Writing C:\Users\\AppData\Local\Temp\tmp5E3F.tmp
info : Adding PackageReference for package 'Altcover' into project 'C:\dev\fxunit\fxunit.fsproj'.
info : Restoring packages for C:\dev\fxunit\fxunit.fsproj...
info :   CACHE https://api.nuget.org/v3-flatcontainer/altcover/index.json
info : Package 'Altcover' is compatible with all the specified frameworks in project 'C:\dev\fxunit\fxunit.fsproj'.
info : PackageReference for package 'Altcover' version '6.2.714' added to file 'C:\dev\fxunit\fxunit.fsproj'.
info : Committing restore...
info : Generating MSBuild file C:\dev\fxunit\obj\fxunit.fsproj.nuget.g.props.
info : Generating MSBuild file C:\dev\fxunit\obj\fxunit.fsproj.nuget.g.targets.
info : Writing assets file to disk. Path: C:\dev\fxunit\obj\project.assets.json
log  : Restore completed in 376.4 ms for C:\dev\fxunit\fxunit.fsproj.
PS C:\dev\fxunit> dotnet build
Microsoft (R) Build Engine version 16.3.0-preview-19426-01+faf5e5d75 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 309.6 ms for C:\dev\fxunit\fxunit.fsproj.
  You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview
  fxunit -> C:\dev\fxunit\bin\Debug\netcoreapp3.0\fxunit.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:03.33
PS C:\dev\fxunit> dotnet test
Test run for C:\dev\fxunit\bin\Debug\netcoreapp3.0\fxunit.dll(.NETCoreApp,Version=v3.0)
Microsoft (R) Test Execution Command Line Tool Version 16.3.0-preview-20190820-01
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...

A total of 1 test files matched the specified pattern.

Test Run Successful.
Total tests: 1
     Passed: 1
 Total time: 1.7223 Seconds
PS C:\dev\fxunit> dotnet test /p:Altcover=true
Test run for C:\dev\fxunit\bin\Debug\netcoreapp3.0\fxunit.dll(.NETCoreApp,Version=v3.0)
Microsoft (R) Test Execution Command Line Tool Version 16.3.0-preview-20190820-01
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...

A total of 1 test files matched the specified pattern.
The active test run was aborted. Reason: Test host process crashed : Stack overflow.


Test Run Aborted.
C:\Users\\.nuget\packages\altcover\6.2.714\build\netstandard2.0\AltCover.targets(115,5): warning : A total of 0 visits recorded [C:\dev\fxunit\fxunit.fsproj]
```

If I do the same thing for a C# xunit project (dotnet new xunit), testing with coverage completes successfully (dotnet test /p:Altcover=true) and generates coverage.xml (16.0 MB for a single dummy assertion, lol).

So something weird is up when combining F# and Altcover.

... Oh I'm using a preview build of dotnet core.  I wonder whether that could be causing...

Yeah.  It's perfectly find on the latest core 2.2.

I wonder why the failure mode of this interaction is a stack overflow, and whether that will be a problem with the core 3.0 release version.

**Comment by SteveGilham** [comment]
Also reproduces with RC1 `Command Line Tool Version 16.3.0-preview-20190905-01`

**Comment by SteveGilham** [comment]
Adding `p:AltCoverAssemblyExcludeFilter=FSharp` to the command line lets the tests run through properly.  Clearly, I'm breaking something in this version of FSharp.Core; but there's at least a work-round for using 3.0 for the moment.

**Comment by SteveGilham** [comment]
Problem diagnosis : when the instrumentation call is made, the recorder assembly calls into the instrumented FSharp.Core, which then calls into the recorder assembly...

The underlying cause is a change in behaviour with .net core 3.0, in which FSharp.Core (and a lot more beside) is being copied into the output folder; it seems to be doing the equivalent of `dotnet publish` (or behaving like the .net Framework always used to).  I work around this in the Framework/Mono build by `ILMerge`ing the FSharp.Core dependency into the recorder, haven't yet managed to get any equivalent solution running in .net core

**Comment by SteveGilham** [comment]
Believed fixed in [release v6.2.719](https://www.nuget.org/packages/AltCover/6.2.719)

**Comment by SteveGilham** [comment]
Something weird going on here; installing the nuget from a local build works just fine; installing the new 6.2.719 nuget still gives a stack overflow.

**Comment by SteveGilham** [comment]
As you were.  The failure was due to the fact that the `fxunit` project had been added to a solution which I'd used to build a debuggable repro in a second project, comparing the two as I went -- and a `dotnet test` was running through `fxunit` OK then borking at the known-to-fail case.  

So I can now say with confidence -- 

Fixed in [release v6.2.719](https://www.nuget.org/packages/AltCover/6.2.719)

**Comment by eternaldensity** [comment]
Fascinating.  I'm glad I reported this!
